File: glyph-mi/python/glyph_mi/confidence.py
"""Confidence scoring for tag suggestions."""


def build_confidence(reasons: list[str], fields: dict) -> dict:
    score = 0
    unique = list(dict.fromkeys(reasons))

    weights = {
        "artist–title filename pattern": 35,
        "artist–album–title pattern": 40,
        "existing artist tag": 25,
        "existing title tag": 20,
        "existing album tag": 15,
        "folder structure": 20,
        "folder hint": 12,
        "album consensus": 30,
        "knowledge pack match": 28,
        "knowledge:": 22,
    }

    for r in unique:
        for key, w in weights.items():
            if key in r:
                score += w
                break
        else:
            if "cleanup" in r:
                score += 8

    filled = sum(1 for k in ("title", "artist", "album") if fields.get(k))
    score += filled * 8
    score = min(100, score)

    if score >= 75:
        level = "high"
    elif score >= 45:
        level = "medium"
    else:
        level = "low"

    return {"score": score, "level": level, "reasons": unique[:12]}

File: glyph-mi/js/core/confidence.js
export function buildConfidence(reasons, fields) {
  let score = 20;
  const hasArtist = Boolean(fields.artist);
  const hasTitle = Boolean(fields.title);
  const hasAlbum = Boolean(fields.album);

  if (hasTitle) score += 25;
  if (hasArtist) score += 30;
  if (hasAlbum) score += 15;
  if (fields.year) score += 5;
  if (fields.trackNo) score += 5;

  const highSignals = reasons.filter((r) =>
    /pattern|folder|consensus|tags|multi-artist/i.test(r)
  ).length;
  score += Math.min(20, highSignals * 6);

  if (reasons.some((r) => /cleanup only|low/i.test(r))) score = Math.min(score, 35);

  score = Math.max(0, Math.min(100, score));

  let level = 'low';
  if (score >= 72) level = 'high';
  else if (score >= 48) level = 'medium';

  return { level, score, reasons: [...reasons] };
}
